### ISSUE TITLE
Property Injection/resolution for @ConfigurationProperties does not honor any registered ProtocolResolver's

### ISSUE THREAD

**Issue by aaronjwhiteside** [spring-projects/spring-boot#11659]
Not sure if this should be a question or a bug report (I could be doing it wrong).

Spring Boot 1.5.9, Spring Framework 4.3.11, JDK 8

I have a custom `ProtocolResolver` registered with:

```java
    public class ConsulProtocolResolverInitializer implements PriorityOrdered, BeanFactoryPostProcessor,
            ResourceLoaderAware {

        private final ConsulClient consulClient;
        private final ConsulConfigProperties configProperties;
        private final Environment env;

        public ConsulProtocolResolverInitializer(final ConsulClient consulClient,
                                                 final ConsulConfigProperties configProperties,
                                                 final Environment env) {
            this.consulClient = consulClient;
            this.configProperties = configProperties;
            this.env = env;
        }

        @Override
        public int getOrder() {
            return Ordered.HIGHEST_PRECEDENCE;
        }

        @Override
        public void postProcessBeanFactory(final ConfigurableListableBeanFactory beanFactory) throws BeansException {
            // we need to implement BeanFactoryPostProcessor + PriorityOrdered to get instantiated early enough to
            // add our custom protocol resolver
        }

        @Override
        public void setResourceLoader(final ResourceLoader resourceLoader) {
            if (resourceLoader instanceof DefaultResourceLoader) {
                ((DefaultResourceLoader)resourceLoader)
                        .addProtocolResolver(new ConsulProtocolResolver(consulClient, configProperties, env));
            }
        }
    }
```

And this works for `@Component` injection + `@Value` annotation:
```java
    @Autowired
    private TestBean testBean;

    @Component
    static class TestBean {
        @Value("${testBean.resource:#{null}}")
        private Resource resource;
    }
```

However it does not work for:
```java
    @Autowired
    private TestBean testBean;

    @ConfigurationProperties("testBean")
    static class TestBean {
        private Resource resource;
   }
```

It seems that my custom `ProtocolResolver` is never invoked for `@ConfigurationProperties` but is for `@Value`..

```
java.lang.IllegalStateException: Failed to load ApplicationContext

	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:124)
	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:83)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.injectDependencies(DependencyInjectionTestExecutionListener.java:117)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.prepareTestInstance(DependencyInjectionTestExecutionListener.java:83)
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:230)
	at org.springframework.test.context.testng.AbstractTestNGSpringContextTests.springTestContextPrepareTestInstance(AbstractTestNGSpringContextTests.java:149)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:108)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:523)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:224)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:146)
	at org.testng.internal.TestMethodWorker.invokeBeforeClassMethods(TestMethodWorker.java:166)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:105)
	at org.testng.TestRunner.privateRun(TestRunner.java:744)
	at org.testng.TestRunner.run(TestRunner.java:602)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:380)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:375)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1301)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1226)
	at org.testng.TestNG.runSuites(TestNG.java:1144)
	at org.testng.TestNG.run(TestNG.java:1115)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:72)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:123)
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'testBean-com.yapstone.rest.framework.springcloud.io.consul.ConsulResourceIT$TestBean': Could not bind properties to ConsulResourceIT.TestBean (prefix=testBean, ignoreInvalidFields=false, ignoreUnknownFields=true, ignoreNestedProperties=false); nested exception is org.springframework.validation.BindException: org.springframework.boot.bind.RelaxedDataBinder$RelaxedBeanPropertyBindingResult: 2 errors
Field error in object 'testBean' on field 'inputStream': rejected value [consul://test.txt]; codes [typeMismatch.testBean.inputStream,typeMismatch.inputStream,typeMismatch.java.io.InputStream,typeMismatch]; arguments [org.springframework.context.support.DefaultMessageSourceResolvable: codes [testBean.inputStream,inputStream]; arguments []; default message [inputStream]]; default message [Failed to convert property value of type 'java.lang.String' to required type 'java.io.InputStream' for property 'inputStream'; nested exception is java.lang.IllegalArgumentException: Failed to retrieve InputStream for class path resource [consul://test.txt]]
Field error in object 'testBean' on field 'path': rejected value [consul://test.txt]; codes [typeMismatch.testBean.path,typeMismatch.path,typeMismatch.java.nio.file.Path,typeMismatch]; arguments [org.springframework.context.support.DefaultMessageSourceResolvable: codes [testBean.path,path]; arguments []; default message [path]]; default message [Failed to convert property value of type 'java.lang.String' to required type 'java.nio.file.Path' for property 'path'; nested exception is java.lang.IllegalArgumentException: Failed to retrieve file for class path resource [consul://test.txt]]
	at org.springframework.boot.context.properties.ConfigurationPropertiesBindingPostProcessor.postProcessBeforeInitialization(ConfigurationPropertiesBindingPostProcessor.java:335)
	at org.springframework.boot.context.properties.ConfigurationPropertiesBindingPostProcessor.postProcessBeforeInitialization(ConfigurationPropertiesBindingPostProcessor.java:292)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:409)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1620)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:761)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:867)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:543)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:693)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:360)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:303)
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:120)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:98)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:116)
	... 29 more
Caused by: org.springframework.validation.BindException: org.springframework.boot.bind.RelaxedDataBinder$RelaxedBeanPropertyBindingResult: 2 errors
Field error in object 'testBean' on field 'inputStream': rejected value [consul://test.txt]; codes [typeMismatch.testBean.inputStream,typeMismatch.inputStream,typeMismatch.java.io.InputStream,typeMismatch]; arguments [org.springframework.context.support.DefaultMessageSourceResolvable: codes [testBean.inputStream,inputStream]; arguments []; default message [inputStream]]; default message [Failed to convert property value of type 'java.lang.String' to required type 'java.io.InputStream' for property 'inputStream'; nested exception is java.lang.IllegalArgumentException: Failed to retrieve InputStream for class path resource [consul://test.txt]]
Field error in object 'testBean' on field 'path': rejected value [consul://test.txt]; codes [typeMismatch.testBean.path,typeMismatch.path,typeMismatch.java.nio.file.Path,typeMismatch]; arguments [org.springframework.context.support.DefaultMessageSourceResolvable: codes [testBean.path,path]; arguments []; default message [path]]; default message [Failed to convert property value of type 'java.lang.String' to required type 'java.nio.file.Path' for property 'path'; nested exception is java.lang.IllegalArgumentException: Failed to retrieve file for class path resource [consul://test.txt]]
	at org.springframework.boot.bind.PropertiesConfigurationFactory.checkForBindingErrors(PropertiesConfigurationFactory.java:359)
	at org.springframework.boot.bind.PropertiesConfigurationFactory.doBindPropertiesToTarget(PropertiesConfigurationFactory.java:276)
	at org.springframework.boot.bind.PropertiesConfigurationFactory.bindPropertiesToTarget(PropertiesConfigurationFactory.java:240)
	at org.springframework.boot.context.properties.ConfigurationPropertiesBindingPostProcessor.postProcessBeforeInitialization(ConfigurationPropertiesBindingPostProcessor.java:330)
	... 47 more

```

**Comment by snicoll** [comment]
You are correct, the binder does not use any of that stuff. That's an interesting report, thanks.

**Comment by aaronjwhiteside** [comment]
Is there a possible workaround? 

Spring's Resource Loader support is awesome, but not being able to customize it when using Spring Boot seems like a pretty big deal.

I would seem that all of the Spring Cloud Config providers are limited by this, as they provide support for protocols like s3:// and gs:// out of the box. But they cannot be applied to most things in Spring Boot (lots of places in Spring Boot where `@ConfigurationProperties` is used internally).

**Comment by snicoll** [comment]
> Is there a possible workaround?

Yes, do not map them as `Resource` (but `String` and then resolve to `Resource` programmatically (you can inject a `ResourceLoader` and call `getResource` on  it).

**Comment by aaronjwhiteside** [comment]
Yeah, I can only do that for things I control/my own code.

Do you have any hints about a possible solution/direction? I'd be happy to investigate and submit a PR..

**Comment by snicoll** [comment]
Fixed by 3db5c70
